File: UITextSubClass/UITextFieldWithNumberPad/UITextFieldWithNumberPad.m
//
//  UITextFieldWithNumberPad.m
//  MotherChildHandbook
//
//  Created by P.I.akura on 2013/09/30.
//  Copyright (c) 2013年 P.I.akura. All rights reserved.
//

#import "UITextFieldWithNumberPad.h"

NS_ASSUME_NONNULL_BEGIN

@implementation UITextFieldWithNumberPad

- (instancetype)init {
    self = [super init];
    if (self) {
        [self setup];
    }
    return self;
}

- (instancetype)initWithFrame:(CGRect)frame {
    self = [super initWithFrame:frame];
    if (self) {
        [self setup];
    }
    return self;
}

- (instancetype _Nullable)initWithCoder:(NSCoder *)aDecoder {
    self = [super initWithCoder:aDecoder];
    if (self) {
        [self setup];
    }
    return self;
}

- (void)setup {
    self.keyboardType = UIKeyboardTypeNumberPad;
    self.menuHidden = YES;
    [self addTarget:self action:@selector(didChangeText:) forControlEvents:UIControlEventEditingChanged];
}

- (void)didChangeText:(id)sender {
    self.value = self.text.integerValue;
}

- (NSInteger)value {
    return [self.text integerValue];
}

- (void)setValue:(NSInteger)value {
    super.text = [@(value) stringValue];
}

- (BOOL)canPerformAction:(SEL)action withSender:(id _Nullable)sender {
    if (self.menuHidden) {
        return NO;
    } else {
        return [super canPerformAction:action withSender:sender];
    }
}

- (NSUndoManager * _Nullable)undoManager {
    return nil;
}

@end

NS_ASSUME_NONNULL_END

File: UITextSubClass/UITextFieldWithNumberPad/UITextFieldWithNumberPad.swift
//
//  UITextFieldWithNumberPad.swift
//  UITextSubClass
//
//  Created by akuraru on 2018/06/12.
//

import Foundation
import UIKit

@objc class UITextFieldWithNumberPad: UITextField {
    private var _value: Int = 0
    public let doneButton = UIBarButtonItem(barButtonSystemItem: .done, target: nil, action: nil)

    public var value: Int {
        get {
            return _value
        }
        set(v) {
            _value = v
            text = String(v)
        }
    }

    func setup() {
        value = 0
        addTarget(self, action: #selector(UITextFieldWithNumberPad.didChangeText), for: .editingChanged)
    }

    @objc func didChangeText() {
        let value = Int(text ?? "")
        if let value = value {
            _value = value
        } else {
            self.value = _value
        }
    }

    override var inputAccessoryView: UIView? {
        get {
            let toolbar = UIToolbar()
            toolbar.isTranslucent = true
            toolbar.sizeToFit()

            let centerSpace = UIBarButtonItem(barButtonSystemItem: .flexibleSpace, target: nil, action: nil)

            toolbar.items = [centerSpace, doneButton]

            return toolbar
        }
        set {
        }
    }

    override var keyboardType: UIKeyboardType {
        get {
            return .numberPad
        }
        set {
        }
    }

    override var undoManager: UndoManager? {
        return nil
    }

    override func canPerformAction(_ action: Selector, withSender sender: Any?) -> Bool {
        return false
    }

    override init(frame: CGRect) {
        super.init(frame: frame)
        setup()
    }

    required init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)
        setup()
    }
}
